analytic_engine/src/role_table/mod.rs

```diff
@@ -0,0 +1,242 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicU8, Ordering},
+        Arc,
+    },
+};
+
+use async_trait::async_trait;
+use common_util::define_result;
+use snafu::Snafu;
+use table_engine::table::{AlterSchemaRequest, WriteRequest};
+
+use crate::{
+    instance::{
+        alter::TableAlterSchemaPolicy,
+        flush_compaction::{TableFlushOptions, TableFlushPolicy},
+        write::TableWritePolicy,
+        write_worker::WorkerLocal,
+        Instance,
+    },
+    space::SpaceRef,
+    table::data::TableDataRef,
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to write table, err:{}", source))]
+    WriteTable {
+        source: Box<dyn std::error::Error + Send + Sync + 'static>,
+    },
+}
+
+define_result!(Error);
+
+/// Those methods are expected to be called by [Instance]
+#[async_trait]
+pub trait RoleTable {
+    fn check_state(&self) -> bool;
+
+    async fn change_role(&self) -> Result<()>;
+
+    // async fn close(&self) -> Result<()>;
+
+    async fn write(
+        &self,
+        request: WriteRequest,
+        instance: &Arc<Instance>,
+        space: &SpaceRef,
+        worker_local: &mut WorkerLocal,
+    ) -> Result<usize>;
+
+    // async fn read(&self, request: ReadRequest) -> Result<PartitionedStreams>;
+
+    async fn flush(
+        &self,
+        flush_opts: TableFlushOptions,
+        instance: &Arc<Instance>,
+        worker_local: &mut WorkerLocal,
+    ) -> Result<()>;
+
+    async fn alter_schema(
+        &self,
+        instance: &Arc<Instance>,
+        worker_local: &mut WorkerLocal,
+        request: AlterSchemaRequest,
+    ) -> Result<()>;
+
+    async fn alter_options(
+        &self,
+        instance: &Arc<Instance>,
+        worker_local: &mut WorkerLocal,
+        options: HashMap<String, String>,
+    ) -> Result<()>;
+}
+
+#[allow(dead_code)]
+#[repr(u8)]
+pub enum TableRole {
+    Invalid = 0,
+    Leader = 1,
+    InSync = 2,
+    NoSync = 3,
+}
+
+pub struct LeaderTable {
+    inner: Arc<LeaderTableInner>,
+}
+
+impl Drop for LeaderTable {
+    fn drop(&mut self) {
+        if Arc::strong_count(&self.inner) == 1 {
+            todo!("notify the state is completely changed")
+        }
+    }
+}
+
+struct LeaderTableInner {
```

Comment:
It seems that `LeaderTableInner` is the actual `LeaderTable` and the current `LeaderTable` should be called `LeaderTableRef` which focuses on reference counting.

Reply:
Not like other `T` & `TRef` pattern in our codebase, here `LeaderTableInner` is not public, others can only hold `LeaderTable`. And type alias `type LeaderTableRef = Arc<LeaderTable>` is not suitable since I customized the `Drop` fn for `LeaderTable`